vxfw/layout/flex.go

```diff
@@ -0,0 +1,107 @@
+package layout
+
+import (
+	"math"
+
+	"git.sr.ht/~rockorager/vaxis"
+	"git.sr.ht/~rockorager/vaxis/vxfw"
+)
+
+// FlexItem is a widget used in a [FlexRow] or [FlexColumn]
+type FlexItem struct {
+	Widget vxfw.Widget
+
+	// Flex determines how much space the widget takes in the layout.
+	// Flex of 0 means the widget will be its inherent size.
+	// Remaining space is divided proportionally to all FlexItems with Flex > 0
+	Flex uint8
+}
+
+// FlexRow is a Widget that lays out its children in a row
+type FlexRow struct {
+	Children []*FlexItem
+}
+
+var _ vxfw.Widget = (*FlexRow)(nil)
+
+func (w *FlexRow) HandleEvent(_ vaxis.Event, _ vxfw.EventPhase) (vxfw.Command, error) {
+	return nil, nil
+}
+
+func (w *FlexRow) Draw(ctx vxfw.DrawContext) (vxfw.Surface, error) {
+	// the width of all children in the first pass, this is their inherent size
+	first_pass_w := uint16(0)
+
+	// number of "flex widths" opted in by the flex items
+	total_flex := uint16(0)
+
+	sizes := make([]uint16, len(w.Children))
+
+	lctx := vxfw.DrawContext{
+		Max:        vxfw.Size{Width: math.MaxUint16, Height: ctx.Max.Height},
+		Characters: ctx.Characters,
+	}
+
+	// Iterate over each child, draw it, and measure the width
+	for i, child := range w.Children {
+		surface, err := child.Widget.Draw(lctx)
+		if err != nil {
+			return vxfw.Surface{}, err
+		}
+
+		first_pass_w += surface.Size.Width
+		total_flex += uint16(child.Flex)
+		sizes[i] = surface.Size.Width
+	}
+
+	children := make([]vxfw.SubSurface, len(w.Children))
+	second_pass_w := uint16(0)
+	max_height := uint16(0)
+
+	// extra space left over that needs to be distributed
+	remaining := ctx.Max.Width - first_pass_w
+
+	for i, child := range w.Children {
+		inherent_w := sizes[i]
+		child_w := uint16(0)
+
+		if child.Flex == 0 {
+			child_w = inherent_w
+		} else if i == len(w.Children)-1 {
+			// last child gets the remaining space
+			child_w = ctx.Max.Width - second_pass_w
+		} else {
+			child_w = inherent_w + (remaining*uint16(child.Flex))/total_flex
+		}
+
+		childctx := vxfw.DrawContext{
+			Min:        vxfw.Size{Width: child_w, Height: 0},
+			Max:        vxfw.Size{Width: child_w, Height: ctx.Max.Height},
+			Characters: ctx.Characters,
+		}
+
+		surface, err := child.Widget.Draw(childctx)
+		if err != nil {
+			return vxfw.Surface{}, err
+		}
+
+		children[i] = vxfw.SubSurface{
+			Origin: vxfw.RelativePoint{
+				Col: int(second_pass_w), Row: 0,
+			},
+			Surface: surface,
+			ZIndex:  0,
+		}
+		if surface.Size.Height > max_height {
+			max_height = surface.Size.Height
+		}
+		second_pass_w += surface.Size.Width
```

Comment:
If I instead put `second_pass_w += child_w` the origins are correct in the test, but the surface widths are still wrong, which is a strong indicator that `text.Draw` is returning a surface that only takes exactly how much it needs.